### PR TITLE
Denoising Autoencoder on MNIST

### PR DESCRIPTION
**Pull request by roatienza**
Adding the classic denoising autoencoder using MNIST dataset.

Sample output:
image

### CODE REVIEW

File: examples/mnist_denoising_autoencoder.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+x = Reshape((shape[1], shape[2], shape[3]))(x)
+
+# Stack of BN-ReLU-Transposed Conv2D-UpSampling blocks
+for i in range(2):
+    x = BatchNormalization()(x)
+    x = Activation('relu')(x)
+    x = Conv2DTranspose(filters=filters,
+                        kernel_size=kernel_size,
+                        padding='same')(x)
+    x = UpSampling2D()(x)
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: You don't need this, `Conv2DTranspose` with strides should work better

File: examples/mnist_denoising_autoencoder.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+# Instantiate Encoder model
+encoder = Model(inputs, latent, name='encoder')
+encoder.summary()
+
+# Build the Decoder model
+latent_inputs = Input(shape=(latent_dim,), name='decoder_input')
+x = Dense(shape[1] * shape[2] * shape[3])(latent_inputs)
+x = Reshape((shape[1], shape[2], shape[3]))(x)
+
+# Stack of BN-ReLU-Transposed Conv2D-UpSampling blocks
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: You don't need BN for such a shallow network, `Conv2DTranspose` with relu activation and strides should suffice

File: examples/mnist_denoising_autoencoder.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+kernel_size = 3
+pool_size = 2
+filters = 16
+latent_dim = 16
+
+# Build the Autoencoder model
+# First build the Encoder model
+inputs = Input(shape=input_shape, name='encoder_input')
+x = inputs
+# Stack of BN-ReLU-Conv2D-MaxPooling blocks
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: You don't need BN for such a shallow network, `Conv2D` and `MaxPooling2D` should suffice

File: examples/mnist_denoising_autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    x = BatchNormalization()(x)
+    x = Activation('relu')(x)
+    filters = filters * 2
+    x = Conv2D(filters=filters,
+               kernel_size=kernel_size,
+               padding='same')(x)
+    x = MaxPooling2D()(x)
+
+# Shape info needed to build decoder model
+shape = x.shape.as_list()
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: This is TensorFlow specific. Instead you can use `K.int_shape(x)`

File: examples/mnist_denoising_autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+'''Trains a denoising autoenconder on MNIST dataset.
+
+Gets val_loss=0.02 after 1 epoch.
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: This is not useful information. Please replace with a few sentences about what is a denoising autoencoder and how it's setup.

Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. More informative documentation added.

File: examples/mnist_denoising_autoencoder.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+Given a training dataset of corrupted data as input and 
+true signal as output, a denoising autoencoder can recover the 
+hidden structure to generate clean data.
+
+This example has modular design. The encoder, decoder and autoencoder
+are 3 models that share weights. For example, after training the 
+autoencoder, the encoder can be used to  generate latent vectors
+of input data for low-dim visualization like PCA or TSNE.
+'''
+
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: Add imports to make the script compatible with py2/3:

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

File: examples/mnist_denoising_autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+# 1) Use Batch Normalization before ReLU on deep networks
+# 2) Use UpSampling2D as alternative to strides>1 
+# - faster but not as good as strides>1
+for i in range(2):
+    x = Activation('relu')(x)
+    x = Conv2DTranspose(filters=filters,
+                        kernel_size=kernel_size,
+                        strides=2,
+                        padding='same')(x)
+    filters = int(filters / 2)
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: filters //= 2

File: examples/mnist_denoising_autoencoder.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+x = Dense(shape[1] * shape[2] * shape[3])(latent_inputs)
+x = Reshape((shape[1], shape[2], shape[3]))(x)
+
+# Stack of ReLU-Transposed Conv2D blocks
+# Notes: 
+# 1) Use Batch Normalization before ReLU on deep networks
+# 2) Use UpSampling2D as alternative to strides>1 
+# - faster but not as good as strides>1
+for i in range(2):
+    x = Activation('relu')(x)
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: Better to put the activation as the `activation` keyword of the layer below

File: examples/mnist_denoising_autoencoder.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+# First build the Encoder Model
+inputs = Input(shape=input_shape, name='encoder_input')
+x = inputs
+# Stack of ReLU-Conv2D blocks
+# Notes: 
+# 1) Use Batch Normalization before ReLU on deep networks
+# 2) Use MaxPooling2D as alternative to strides>1 
+# - faster but not as good as strides>1
+for i in range(2):
+    x = Activation('relu')(x)
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: Better to put the activation as the `activation` keyword of the layer below

File: examples/mnist_denoising_autoencoder.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    filters = filters * 2
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: filters *= 2

File: examples/mnist_denoising_autoencoder.py (hunk 3 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: I was thinking if the end user decided to scale up, the building blocks are already present. I removed the BN and placed it on the comment. MaxPooling2D is also replaced by strides since it is better (though slower) than strides>1.

File: examples/mnist_denoising_autoencoder.py (hunk 4 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. Thanks.

File: examples/mnist_denoising_autoencoder.py (hunk 2 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. However, for deep networks UpSampling2D is faster (though not as good as strides>1). I indicated this as a note.

File: examples/mnist_denoising_autoencoder.py (hunk 1 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Thanks. I included comments on the advantage/disadvantage of UpSampling2D.

File: examples/mnist_denoising_autoencoder.py (hunk 6 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. Thanks.

File: examples/mnist_denoising_autoencoder.py (hunk 9 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. Thanks.

File: examples/mnist_denoising_autoencoder.py (hunk 10 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. Thanks.

File: examples/mnist_denoising_autoencoder.py (hunk 8 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. Thanks.

File: examples/mnist_denoising_autoencoder.py (hunk 7 above)
Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: Fixed. Thanks.

File: examples/mnist_denoising_autoencoder.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+# Build the Autoencoder Model
+# First build the Encoder Model
+inputs = Input(shape=input_shape, name='encoder_input')
+x = inputs
+# Stack of Conv2D blocks
+# Notes:
+# 1) Use Batch Normalization before ReLU on deep networks
+# 2) Use MaxPooling2D as alternative to strides>1
+# - faster but not as good as strides>1
+for i in range(2):
```
Comment (fchollet) [inline examples/mnist_denoising_autoencoder.py]: Here we actually start with 32 filters even though we previously defined `filters = 16`. What would you think of instead having:

```python
layer_filters = [32, 64]
for filters in layer_filters:
    ...

for filters in layer_filters[::-1]:
   # decoder
   ...
```
This allows arbitrary layer size and arbitrary number of layers.

Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: In encoder, the doubling of filter size was actually inspired by autoencoders used in practice, by ResNet and by DenseNet (that uses growth rate). As the number of feature maps is halved, the number of filters doubles. 
In decoder, the halving of filter size has the same motivation. 
Use of array is another way to do it. More flexible but requires more care due to manual encoding of hyperparameters.

Comment (fchollet) [reply examples/mnist_denoising_autoencoder.py]: You are manually encoding the hyperparameters in both cases (number of layers and size of each layer).

Comment (roatienza) [reply examples/mnist_denoising_autoencoder.py]: I Agree. Use of list results to a more compact and flexible code. The autoencoder model structure, like depth, is not constrained to halving/doubling of number of filters limitations. I submitted the revised code.
